File: scripts/metrics_report.js
#!/usr/bin/env node
/**
 * Quantova chain metrics report — qweb3.js
 *
 * Collects a read-only performance and identity snapshot from a Quantova endpoint
 * using the official qweb3.js client. It samples twice over a short interval to
 * derive observed block time and finality lag. It signs nothing and submits nothing.
 *
 * Usage:
 *   node metrics_report.js --rpc https://mainnet.quantova.io
 *   node metrics_report.js --rpc https://testnet.quantova.io --interval 10
 *   node metrics_report.js --rpc http://127.0.0.1:9933
 *
 * Requires: npm install qweb3.js
 */

const { QWeb3 } = require('qweb3.js');

function arg(name, def) {
  const i = process.argv.indexOf(`--${name}`);
  return i !== -1 && process.argv[i + 1] ? process.argv[i + 1] : def;
}

const RPC = arg('rpc', 'https://mainnet.quantova.io');
const INTERVAL = parseInt(arg('interval', '8'), 10);

const sleep = (s) => new Promise((r) => setTimeout(r, s * 1000));
const toNum = (v) => (typeof v === 'string' && v.startsWith('0x') ? parseInt(v, 16) : Number(v));

async function main() {
  const q = new QWeb3(RPC);
  console.log(`Quantova metrics report (qweb3.js)`);
  console.log(`Endpoint: ${RPC}`);
  console.log('-'.repeat(60));

  // Identity / runtime
  try {
    const rv = await q.rpc.call('state_getRuntimeVersion', []);
    console.log(`Runtime         : ${rv.specName} v${rv.specVersion} (tx v${rv.transactionVersion})`);
  } catch (e) { console.log(`Runtime         : (unavailable: ${e.message})`); }

  // Sample 1
  const b1 = toNum(await q.rpc.blockNumber());
  let f1 = null;
  try { f1 = toNum((await q.rpc.call('chain_getHeader', [await q.rpc.call('chain_getFinalizedHead', [])])).number); } catch (_) {}
  const t1 = Date.now();

  await sleep(INTERVAL);

  // Sample 2
  const b2 = toNum(await q.rpc.blockNumber());
  let f2 = null;
  try { f2 = toNum((await q.rpc.call('chain_getHeader', [await q.rpc.call('chain_getFinalizedHead', [])])).number); } catch (_) {}
  const t2 = Date.now();

  const elapsed = (t2 - t1) / 1000;
  const blocks = b2 - b1;
  const blockTime = blocks > 0 ? (elapsed / blocks).toFixed(2) : 'n/a';

  console.log(`Best block      : ${b1} -> ${b2}  (+${blocks} in ~${elapsed.toFixed(0)}s)`);
  console.log(`Observed block time: ~${blockTime}s  (target ~2.5s)`);
  if (f2 !== null) {
    console.log(`Finalized head  : ${f1} -> ${f2}`);
    console.log(`Finality lag    : ${b2 - f2} blocks behind best`);
  }

  // Fees
  try {
    const fees = await q.fees.estimate();
    console.log(`Fee tiers       : slow=${fees.tiers.slow} standard=${fees.tiers.standard} fast=${fees.tiers.fast} (${fees.model})`);
  } catch (e) { console.log(`Fee tiers       : (unavailable: ${e.message})`); }

  console.log('-'.repeat(60));
  console.log('Read-only snapshot complete. No transactions were sent.');
}

main().catch((e) => { console.error('error:', e.message); process.exit(1); });

File: scripts/metrics_report.py
#!/usr/bin/env python3
"""Quantova chain metrics report — qweb3.py

Collects a read-only performance and identity snapshot from a Quantova endpoint
using the official qweb3.py client. It samples twice over a short interval to derive
observed block time and finality lag. It signs nothing and submits nothing.

Usage:
    python metrics_report.py --rpc https://mainnet.quantova.io
    python metrics_report.py --rpc https://testnet.quantova.io --interval 10
    python metrics_report.py --rpc http://127.0.0.1:9933

Requires: pip install qweb3
"""

import argparse
import time

from qweb3 import QWeb3


def to_num(v):
    if isinstance(v, str) and v.startswith("0x"):
        return int(v, 16)
    return int(v)


def main():
    p = argparse.ArgumentParser(description="Quantova chain metrics report (qweb3.py)")
    p.add_argument("--rpc", default="https://mainnet.quantova.io", help="JSON-RPC endpoint")
    p.add_argument("--interval", type=int, default=8, help="seconds between samples")
    args = p.parse_args()

    q = QWeb3(args.rpc)
    print("Quantova metrics report (qweb3.py)")
    print(f"Endpoint: {args.rpc}")
    print("-" * 60)

    # Identity / runtime
    try:
        rv = q.rpc.call("state_getRuntimeVersion", [])
        print(f"Runtime            : {rv['specName']} v{rv['specVersion']} (tx v{rv['transactionVersion']})")
    except Exception as e:
        print(f"Runtime            : (unavailable: {e})")

    # Sample 1
    b1 = to_num(q.rpc.block_number())
    f1 = None
    try:
        f1 = to_num(q.rpc.call("chain_getHeader", [q.rpc.call("chain_getFinalizedHead", [])])["number"])
    except Exception:
        pass
    t1 = time.time()

    time.sleep(args.interval)

    # Sample 2
    b2 = to_num(q.rpc.block_number())
    f2 = None
    try:
        f2 = to_num(q.rpc.call("chain_getHeader", [q.rpc.call("chain_getFinalizedHead", [])])["number"])
    except Exception:
        pass
    t2 = time.time()

    elapsed = t2 - t1
    blocks = b2 - b1
    block_time = f"{elapsed / blocks:.2f}" if blocks > 0 else "n/a"

    print(f"Best block         : {b1} -> {b2}  (+{blocks} in ~{elapsed:.0f}s)")
    print(f"Observed block time: ~{block_time}s  (target ~2.5s)")
    if f2 is not None:
        print(f"Finalized head     : {f1} -> {f2}")
        print(f"Finality lag       : {b2 - f2} blocks behind best")

    # Fees
    try:
        fees = q.fees.estimate()
        t = fees["tiers"]
        print(f"Fee tiers          : slow={t['slow']} standard={t['standard']} fast={t['fast']} ({fees.get('model')})")
    except Exception as e:
        print(f"Fee tiers          : (unavailable: {e})")

    print("-" * 60)
    print("Read-only snapshot complete. No transactions were sent.")


if __name__ == "__main__":
    main()
